addons/product/models/product_template.py

```diff
@@ -502,6 +543,226 @@ def has_dynamic_attributes(self):
+                        lambda product_attribute_value: product_attribute_value.attribute_id.create_variant == 'no_variant'
+                    )
+                ]
+            )
+            product_id = product.id
+            list_price = product.price_compute('list_price')[product.id]
+            price = product.price or list_price
+
+        display_name = [product_template.name]
+        if filtered_product_template_attribute_values:
+            display_name.append(' (')
+            display_name.append(', '.join(filtered_product_template_attribute_values.mapped('name')))
+            display_name.append(')')
+
+        if pricelist and pricelist.currency_id != product_template.currency_id:
+            list_price = product_template.currency_id._convert(
+                list_price, pricelist.currency_id,
+                self.env.user.company_id, fields.Date.today()
+            )
+
+        return {
+            'product_id': product_id,
+            'product_template_id': product_template.id,
+            'display_name': ''.join(display_name),
+            'price': price,
+            'list_price': list_price
+        }
+
+    @api.multi
+    def is_combination_possible(self, combination, reference_product=None):
+        """
+            :param combination: recordset of product.template.attribute.value
+            :param reference_product: `product.product` from which 'self' is an
+                optional or accessory product
+            :return: boolean: True if the combination is possible
+        """
+        self.ensure_one()
+        combinationData = self._get_attribute_exclusions(reference_product=None)
+        exclusions = combinationData.get('exclusions')
+        parent_exclusions = combinationData.get('parent_exclusions')
+        archived_combinations = combinationData.get('archived_combinations')
+
+        disable = False
+
+        if exclusions:
+            # exclude if the current value is in an exclusion,
+            # and the value excluding it is also in the combination
+            for product_template_attribute_value_id in combination.ids:
+                if product_template_attribute_value_id in exclusions:
+                    for exclusion in exclusions.get(product_template_attribute_value_id):
+                        if exclusion in combination.ids:
+                            disable = True
+
+        for exclusion in parent_exclusions:
+            if exclusion in combination.ids:
+                disable = True
+
+        for archived_combination in archived_combinations:
+            # ignore the archived combination if it has a different number of attributes
+            if len(archived_combination) != len(combination):
+                break
+
+            for attribute_value in archived_combination:
+                if attribute_value not in combination.ids:
+                    disable = True
+
+        return not disable
+
+    @api.multi
+    def get_first_possible_combination(self):
```

Comment:
@seb-odoo 

Should these new methods be public? Is there any good reason?

Reply:
This is a good concern, and actually it depends which method.

In general those that are called from a route should be public so we can make the RPC directly on them instead and get rid of the routes.

The others should probably be made private.

I'll clean this up at the end.

Reply:
Every added method should be private now.